src/qt3utils/datagenerators/princeton.py

```diff
@@ -0,0 +1,339 @@
+        self._experiment = self._application.Experiment
+
+    @property
+    def automation(self):
+        return self._automation
+
+    @property
+    def addinbase(self):
+        return self._addinbase
+
+    @property
+    def application(self):
+        return self._application
+
+    @property
+    def experiment(self):
+        return self._experiment
+
+    def __del__(self):
+        """
+        Uses Python garbage collection method used to terminate
+        AddInProcess.exe, if this is not done, LightField will not reopen
+        properly.
+        """
+        self.automation.Dispose()
+        print('Closed AddInProcess.exe')
+
+    def set(self, setting, value):
+        """
+        Helper function for setting experiment parameters with basic value
+        checking.
+        """
+
+        if self.experiment.Exists(setting):
+            if self.experiment.IsValid(setting, value):
+
+                self.experiment.SetValue(setting, value)
+
+            else:
+
+                # TODO: might need to add a proper exception
+                print('Invalid value: {} for setting: {}.'.format(value, setting))
+
+        else:
+
+            # TODO: might need to add a proper exception
+            print('Invalid setting:{}'.format(setting))
+
+        return
+
+    def get(self, setting):
+        """
+        Helper function for getting experiment parameters with a basic check
+        to see if a specific setting is valid.
+        """
+
+        if self.experiment.Exists(setting):
+
+            value = self.experiment.GetValue(setting)
+
+        else:
+
+            value = []
+            print('Invalid setting: {}'.format(setting))
+
+        return value
+
+    def load_experiment(self, value):
+        self.experiment.Load(value)
+
+    def acquire(self):
+        """
+        Helper function to acquire the data from the PIXIS-style cameras.
+        Need to check array reorganization since there is only a single vertical pixel.
+        For a single frame, the data is returned as a 2D array (just raw counts
+        from each pixel). Taking a section in the vertical (400-pixel) direction
+        corresponds to wavelength averaging.
+        For multiple frames, each exposure is stacked in a third dimension, so
+        averaging can be performed simply by summing over the different planes.
+        """
+
+        acquisition_time = self.get(lf.AddIns.CameraSettings.ShutterTimingExposureTime) #to allow the camera to capture the desired amount of data during the specified exposure time.
+        num_frames = self.get(lf.AddIns.ExperimentSettings.FrameSettingsFramesToStore)
+
+        self.experiment.Acquire()
+
+        sleep(0.001 * acquisition_time * num_frames)  #sleep delay that waits for the exposure duration of the camera
+
+        while self.experiment.IsRunning:
+            sleep(0.1)  #loop that repeatedly checks whether the experiment is still running so it decided when to move to next block of code
```

Comment:
Is it possible to get stuck inside an infinite loop here? What happens if the spectrometer loses power? Consider adding a timeout option?  How long should it take to acquire data? Can you estimate it based on the settings in the object? You could then add some reasonable buffer and generate a max timeout to make it impossible to get stuck in infinite loop.

Reply:
No, if it checks and sees that the Spectrometer is off then it will shut the program down. I was thinking of adding a time estimation tool before, but Max and I decided to only add that later down the line. I will look into adding the timeout option as well, the only issue is that the Lightfield software does not like being shutdown prematurely unless it has a firm reason to do so.